draft-ietf-quic-qlog-quic-events.md

```diff
@@ -1958,35 +1965,46 @@ that packet with number 120 was ACKed). However, in that case, implementers SHOU
 log \[120\] instead and tools MUST be able to deal with both notations.
 
 ### ResetStreamFrame
+
+If the error_code numerical value does not map to a known ApplicationError string,
+the error_code value of "unknown" can be used and the raw value captured in the
+error_code_bytes field; a numerical value without variable-length integer
+encoding.
+
 ~~~ cddl
 ResetStreamFrame = {
     frame_type: "reset_stream"
     stream_id: uint64
-    error_code: $ApplicationError /
-                uint64
+    error_code: $ApplicationError
```

Comment:
So for my library, the QUIC layer is responsible for generating qlog frames and it is not application aware. That means I always just send the numerical code right now in `error_code`. Requiring me to send "unknown" will inflate my logs without any real benefit. Can we consider making `error_code` optional if going this route?

Reply:
Gosh, I'm not the biggest fan of this... the main reason to do this type of change (imo) is so it's easier what to look at when parsing a qlog. With this, you can always assume the `error_code` is a string and just use that, and only if the value is `unknown` do you need to maybe look for another field. If you make `error_code` optional and people have to look for the optional presence of `error_code_bytes` as well, then there's not that much benefit to changing this from `$ApplicationError / uint64` to two different fields, imo

Reply:
What do you mean by "look at when parsing"? My parser attempts to parse all the defined fields statelessly. It's autogenerated code that currently can determine the qlog type based on whether its parsing a JSON string or number. Since JSON doesn't require order of fields, to be maintained, I'm not sure how a parser can do much statefully without already having parsed the object.

If you mean tools dealoling with the qlog object, then I can follow that argument. However, I just render the fields that are in the log, rather than do any checking or enforcement.

I can live with the current definition, it's straightforward and avoids weird scenarios where a well known error code is logged simply ad an integer rather than a string claiming "unknown", which is confusing IMO.

Reply:
In that case, I don't think it's too onerous on a more-capable tool to allow for error_code to be optional. It can just treat an omitted field as the default "unknown"

Reply:
another option would be for me to log an empty string (which I think is inherently allowed) - that doesn't require any change to this PR and lets me shave some unnecessary/mislabelled bytes of `"unknown"`

Reply:
If that's ok for you, then that works for me as well :)